### PR TITLE
node: Support custom registries

### LINKED ISSUE #316: flatpak-node-generator fails with sha1 integrity check
### flatpak-builder version

1.2.0

### Linux distribution and version

Ubuntu 20.04

### Affected flatpak-builder tool

node/flatpak-node-generator.py

### flatpak-builder tool cli args

--no-requests-cache --no-xdg-layout

### Source repository URL

https://raw.githubusercontent.com/cinnyapp/cinny/v2.2.0/package-lock.json

### Flatpak-builder manifest URL

https://raw.githubusercontent.com/flathub/in.cinny.Cinny/master/in.cinny.Cinny.yml

### Description

I have updated my dependency (@matrix-org/olm) and it has sha1 integrity hash now. Now builder seems to be failing when processing that dependency to generate node sources.

Logs: https://github.com/flathub/in.cinny.Cinny/actions/runs/3073712103/jobs/4966018497#step:12:3399

### PR DESCRIPTION
**Pull request by gasinvein**
This is a rather quick-and-dirty implementation. We probably can implement custom registries in a nicer way, but at least this does the job.
Fixes #316

### CODE REVIEW

File: node/flatpak_node_generator/providers/npm.py [hunk 1]
```diff
@@ -321,6 +324,29 @@ async def generate_package(self, package: Package) -> None:
     def relative_lockfile_dir(self, lockfile: Path) -> Path:
         return lockfile.parent.relative_to(self.lockfile_root)
 
+    @functools.lru_cache(typed=True)
+    def get_lockfile_rc(self, lockfile: Path) -> Dict[str, str]:
+        rc = {}
+        rcfile_paths = []
+        for path in lockfile.parents:
```
Comment (refi64) [inline node/flatpak_node_generator/providers/npm.py:331]: Afaik npmrc files aren't evaluated up a tree, only yarnrc files?

File: node/flatpak_node_generator/providers/npm.py [hunk 2]
```diff
@@ -142,6 +143,8 @@ def __init__(
         self.git_sources: DefaultDict[
             Path, Dict[Path, GitSource]
         ] = collections.defaultdict(lambda: {})
+        # FIXME better pass the same provider object we created in main
```
Comment (refi64) [inline node/flatpak_node_generator/providers/npm.py:146]: this hurts :cry: but it's tricky to change, I'm playing around a bit with it rn but we can definitely just clean it all up later.

File: node/flatpak_node_generator/providers/npm.py [hunk 3]
```diff
@@ -321,6 +324,29 @@ async def generate_package(self, package: Package) -> None:
...
+                rcfile_paths.append(rcfile_path)
+        for rcfile_path in reversed(rcfile_paths):
+            rc.update(self.rcfile_provider.parse_rcfile(rcfile_path))
+        return rc
+
+    def get_package_registry(self, package: Package) -> str:
+        assert isinstance(package.source, RegistrySource)
+        rc = self.get_lockfile_rc(package.lockfile)
+        if rc and '/' in package.name:
+            org, _ = package.name.split('/', maxsplit=1)
```
Comment (refi64) [inline node/flatpak_node_generator/providers/npm.py:345]:
```suggestion
            scope, _ = package.name.split('/', maxsplit=1)
```

File: node/flatpak_node_generator/providers/npm.py (hunk 1 above)
Comment (gasinvein) [reply node/flatpak_node_generator/providers/npm.py:331]: Not sure. Is it documented somewhere? Or any hints in npm source code?

File: node/flatpak_node_generator/providers/npm.py (hunk 2 above)
Comment (gasinvein) [reply node/flatpak_node_generator/providers/npm.py:146]: It does. TBH while implementing this, I started to doubt if adding a separate factory class for rcfiles was a good idea at all. But ofc this is out of scope of this PR.

File: node/flatpak_node_generator/providers/npm.py (hunk 1 above)
Comment (gasinvein) [reply node/flatpak_node_generator/providers/npm.py]: I'll just take your word on it. Even if it proves wrong, we can put back the tree lookup code.

**Pull request merged**